File: src/monid_finance_mcp/providers/us/statements.py
"""DefiLlama /statements parsing shared by the Financial Datasets tools.

Parses the matrix layout (labels x periodEnding values, with per-parent
children blocks) into per-period value rows keyed by plain statement
labels, derives fiscal-period labels, and derives trailing-twelve-month
windows from consecutive quarters.
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import date

from monid_finance_mcp.models import JsonObject, JsonValue
from monid_finance_mcp.providers.us.normalize import SchemaDriftError

STATEMENT_SECTIONS = {
    "income": ("incomeStatement", "income_statement", "income"),
    "balance": ("balanceSheet", "balance_sheet", "balance"),
    "cash": ("cashflow", "cashFlow", "cash_flow", "cashFlowStatement"),
}


@dataclass(frozen=True, slots=True)
class PeriodRow:
    """One report period of one statement, with plain-label values."""

    report_period: date
    values: dict[str, JsonValue]


@dataclass(frozen=True, slots=True)
class StatementSeries:
    annual: list[PeriodRow]
    quarterly: list[PeriodRow]


def parse_statement_series(value: JsonValue, statement: str) -> StatementSeries:
    """Parse the annual and quarterly matrices of one statement."""
    if not isinstance(value, dict):
        raise SchemaDriftError("DefiLlama statements payload must be an object")
    root: JsonObject = value
    section = _section_root(root, statement)
    return StatementSeries(
        annual=_parse_matrix(section, statement, "annual"),
        quarterly=_parse_matrix(section, statement, "quarterly"),
    )


def fiscal_year_end_month(series: StatementSeries) -> int | None:
    """Learn the fiscal year end month from the annual series."""
    months = {row.report_period.month for row in series.annual}
    return months.pop() if len(months) == 1 else None


def fiscal_period_label(
    row: PeriodRow, year_end_month: int | None, *, is_annual: bool
) -> str | None:
    """Derive the Financial Datasets fiscal_period label for one row."""
    if is_annual:
        return f"FY{row.report_period.year}"
    if year_end_month is None:
        return None
    year = row.report_period.year
    if row.report_period.month > year_end_month:
        year += 1
    quarter = ((row.report_period.month - 1) // 3) + 1
    return f"Q{quarter} FY{year}"


def derive_ttm_rows(
    quarterly: list[PeriodRow], *, flow_labels: frozenset[str], mean_labels: frozenset[str]
) -> list[PeriodRow]:
    """Derive TTM rows from windows of four consecutive quarters.

    Flow labels are summed across the window; mean labels (weighted
    average share counts) are averaged; every other label keeps the
    value from the final quarter of the window (point-in-time balances).
    """
    ordered = sorted(quarterly, key=lambda row: row.report_period)
    rows: list[PeriodRow] = []
    for start in range(len(ordered) - 3):
        window = ordered[start : start + 4]
        if not _consecutive_quarters(window):
            continue
        last = window[3]
        values: dict[str, JsonValue] = {}
        labels: set[str] = set(last.values)
        for row in window:
            labels.update(row.values)
        for label in labels:
            window_values = [row.values.get(label) for row in window]
            numbers = [
                v for v in window_values if isinstance(v, int | float) and not isinstance(v, bool)
            ]
            if len(numbers) != 4:
                continue
            if label in flow_labels:
                values[label] = sum(numbers)
            elif label in mean_labels:
                values[label] = sum(numbers) / 4
            else:
                values[label] = numbers[3]
        rows.append(PeriodRow(report_period=last.report_period, values=values))
    return rows


def filter_rows(
    rows: list[PeriodRow],
    *,
    exact: date | None,
    gte: date | None,
    lte: date | None,
    gt: date | None,
    lt: date | None,
) -> list[PeriodRow]:
    """Apply Financial Datasets report_period comparison filters."""

    def keep(row: PeriodRow) -> bool:
        day = row.report_period
        if exact is not None and day != exact:
            return False
        if gte is not None and day < gte:
            return False
        if lte is not None and day > lte:
            return False
        if gt is not None and day <= gt:
            return False
        return lt is None or day < lt

    return [row for row in rows if keep(row)]


def _consecutive_quarters(window: list[PeriodRow]) -> bool:
    previous: int | None = None
    for row in window:
        ordinal = row.report_period.year * 4 + (row.report_period.month - 1) // 3
        if previous is not None and ordinal != previous + 1:
            return False
        previous = ordinal
    return True


def _section_root(root: JsonObject, statement: str) -> JsonObject:
    for key in STATEMENT_SECTIONS[statement]:
        section = root.get(key)
        if isinstance(section, dict):
            return section
    raise SchemaDriftError(f"DefiLlama payload omitted the {statement} statement")


def _parse_matrix(section: JsonObject, statement: str, period: str) -> list[PeriodRow]:
    block = section.get(period)
    if block is None:
        return []
    if not isinstance(block, dict):
        raise SchemaDriftError(f"DefiLlama {statement}.{period} has an unknown shape")
    labels = _labels(section.get("labels"), f"DefiLlama {statement} labels")
    dates = _dates(block.get("periodEnding"), f"DefiLlama {statement}.{period}.periodEnding")
    values = _value_rows(
        block.get("values"),
        row_count=len(labels),
        column_count=len(dates),
        name=f"DefiLlama {statement}.{period}.values",
    )
    rows = [PeriodRow(report_period=day, values={}) for day in dates]
    for row_index, label in enumerate(labels):
        for column in range(len(dates)):
            item = values[row_index][column]
            if item is not None:
                rows[column].values[label] = item
    children_block = block.get("children")
    child_definitions = section.get("children")
    if isinstance(children_block, dict) and isinstance(child_definitions, dict):
        _parse_children(
            children_block,
            child_definitions,
            period,
            dates,
            rows,
            name=f"DefiLlama {statement}.{period}",
        )
    return rows


def _parse_children(
    children_block: JsonObject,
    child_definitions: JsonObject,
    period: str,
    dates: list[date],
    rows: list[PeriodRow],
    *,
    name: str,
) -> None:
    definitions = child_definitions.get(period)
    if not isinstance(definitions, dict):
        return
    for parent, parent_block in children_block.items():
        if not isinstance(parent_block, dict):
            raise SchemaDriftError(f"{name} children must map parents to blocks")
        definition_value = definitions.get(parent)
        if definition_value is None:
            continue
        if not isinstance(definition_value, dict):
            raise SchemaDriftError(f"{name} child labels for {parent} must be a list")
        definition = definition_value
        child_labels_value = definition.get("labels")
        child_values_value = parent_block.get("values")
        if child_labels_value is None and child_values_value is None:
            continue
        if not isinstance(child_labels_value, list):
            raise SchemaDriftError(f"{name} child labels for {parent} must be a list")
        if not isinstance(child_values_value, list):
            raise SchemaDriftError(f"{name} child values for {parent} must be a list")
        child_labels = [label for label in child_labels_value if isinstance(label, str)]
        child_values = _value_rows(
            child_values_value,
            row_count=len(child_labels),
            column_count=len(dates),
            name=f"{name} children of {parent}",
        )
        for row_index, label in enumerate(child_labels):
            for column in range(len(dates)):
                item = child_values[row_index][column]
                if item is not None:
                    rows[column].values[f"{parent}|{label}"] = item


def _labels(value: JsonValue, name: str) -> list[str]:
    if not isinstance(value, list):
        raise SchemaDriftError(f"{name} must be a list")
    labels: list[str] = []
    for label in value:
        if not isinstance(label, str):
            raise SchemaDriftError(f"{name} must contain strings")
        labels.append(label)
    return labels


def _dates(value: JsonValue, name: str) -> list[date]:
    if not isinstance(value, list):
        raise SchemaDriftError(f"{name} must be a list")
    dates: list[date] = []
    for item in value:
        if not isinstance(item, str):
            raise SchemaDriftError(f"{name} must contain strings")
        try:
            dates.append(date.fromisoformat(item[:10]))
        except ValueError as error:
            raise SchemaDriftError(f"{name} contains a non-ISO date") from error
    return dates


def _value_rows(
    value: JsonValue, *, row_count: int, column_count: int, name: str
) -> list[list[JsonValue]]:
    if not isinstance(value, list) or len(value) != row_count:
        raise SchemaDriftError(f"{name} row count does not match labels")
    rows: list[list[JsonValue]] = []
    for row_index, raw_row in enumerate(value):
        if not isinstance(raw_row, list) or len(raw_row) != column_count:
            raise SchemaDriftError(f"{name} row {row_index} has the wrong width")
        rows.append(list(raw_row))
    return rows

File: go/providers/statements.go
// Package providers holds the DefiLlama-to-Financial-Datasets parsing that
// used to live in monid_finance_mcp/providers/us/*.py. Every function here
// is a direct port of the matching Python module; the Python source is the
// executable spec and every number, label, and key order was verified
// against the live Financial Datasets API.
package providers

import (
	"fmt"
	"sort"
	"time"
)

// dateLayout is the ISO 8601 date-only layout used for report periods
// throughout this package (Python's date.fromisoformat/.isoformat()).
const dateLayout = "2006-01-02"

// statementSections lists the DefiLlama section-key aliases per statement.
// Ported from statements.STATEMENT_SECTIONS.
var statementSections = map[string][]string{
	"income":  {"incomeStatement", "income_statement", "income"},
	"balance": {"balanceSheet", "balance_sheet", "balance"},
	"cash":    {"cashflow", "cashFlow", "cash_flow", "cashFlowStatement"},
}

// PeriodRow is one report period of one statement, with plain-label values.
// Ported from statements.PeriodRow.
type PeriodRow struct {
	ReportPeriod time.Time
	Values       map[string]any
}

// StatementSeries holds the annual and quarterly rows of one statement.
// Ported from statements.StatementSeries.
type StatementSeries struct {
	Annual    []PeriodRow
	Quarterly []PeriodRow
}

// ParseStatementSeries parses the annual and quarterly matrices of one
// statement ("income", "balance", or "cash") from a DefiLlama
// /equities/v1/statements payload. Ported from
// statements.parse_statement_series.
func ParseStatementSeries(value any, statement string) (StatementSeries, error) {
	root, ok := value.(map[string]any)
	if !ok {
		return StatementSeries{}, schemaDriftf("DefiLlama statements payload must be an object")
	}
	section, err := sectionRoot(root, statement)
	if err != nil {
		return StatementSeries{}, err
	}
	annual, err := parseMatrix(section, statement, "annual")
	if err != nil {
		return StatementSeries{}, err
	}
	quarterly, err := parseMatrix(section, statement, "quarterly")
	if err != nil {
		return StatementSeries{}, err
	}
	return StatementSeries{Annual: annual, Quarterly: quarterly}, nil
}

// FiscalYearEndMonth learns the fiscal year end month from the annual
// series. Ported from statements.fiscal_year_end_month.
func FiscalYearEndMonth(series StatementSeries) *int {
	return fiscalYearEndMonthOfRows(series.Annual)
}

func fiscalYearEndMonthOfRows(rows []PeriodRow) *int {
	months := map[int]struct{}{}
	for _, row := range rows {
		months[int(row.ReportPeriod.Month())] = struct{}{}
	}
	if len(months) != 1 {
		return nil
	}
	for m := range months {
		v := m
		return &v
	}
	return nil
}

// FiscalPeriodLabel derives the Financial Datasets fiscal_period label for
// one row: "FY2025" for annual rows, "2026-Q3" (year-first) for quarterly
// rows. Ported from statements.fiscal_period_label.
func FiscalPeriodLabel(row PeriodRow, yearEndMonth *int, isAnnual bool) *string {
	return fiscalPeriodLabel(row.ReportPeriod, yearEndMonth, isAnnual)
}

// fiscalPeriodLabel is the shared implementation used by statements.go,
// metrics.go, and earnings.go alike (all three Python modules compute this
// identically).
func fiscalPeriodLabel(reportPeriod time.Time, yearEndMonth *int, isAnnual bool) *string {
	if isAnnual {
		label := fmt.Sprintf("FY%d", reportPeriod.Year())
		return &label
	}
	if yearEndMonth == nil {
		return nil
	}
	year := reportPeriod.Year()
	if int(reportPeriod.Month()) > *yearEndMonth {
		year++
	}
	quarter := (int(reportPeriod.Month())-1)/3 + 1
	label := fmt.Sprintf("%d-Q%d", year, quarter)
	return &label
}

// quarterOrdinal is a monotonic quarter index, shared across the package.
func quarterOrdinal(t time.Time) int {
	return t.Year()*4 + (int(t.Month())-1)/3
}

// LabelSet is a membership set of statement labels, used to classify TTM
// aggregation: flows are summed, means are averaged, everything else is a
// point-in-time balance carried from the final quarter.
type LabelSet map[string]struct{}

// NewLabelSet builds a LabelSet from the given labels.
func NewLabelSet(labels ...string) LabelSet {
	set := make(LabelSet, len(labels))
	for _, label := range labels {
		set[label] = struct{}{}
	}
	return set
}

func (s LabelSet) has(label string) bool {
	_, ok := s[label]
	return ok
}

// DeriveTTMRows derives TTM rows from windows of four consecutive quarters.
// Flow labels are summed across the window; mean labels (weighted-average
// share counts) are averaged; every other label keeps the value from the
// final quarter of the window (point-in-time balances). Ported from
// statements.derive_ttm_rows.
func DeriveTTMRows(quarterly []PeriodRow, flowLabels, meanLabels LabelSet) []PeriodRow {
	ordered := make([]PeriodRow, len(quarterly))
	copy(ordered, quarterly)
	sort.Slice(ordered, func(i, j int) bool {
		return ordered[i].ReportPeriod.Before(ordered[j].ReportPeriod)
	})

	var rows []PeriodRow
	for start := 0; start+4 <= len(ordered); start++ {
		window := ordered[start : start+4]
		if !consecutiveQuarters(window) {
			continue
		}
		last := window[3]
		labels := map[string]struct{}{}
		for _, row := range window {
			for label := range row.Values {
				labels[label] = struct{}{}
			}
		}
		values := map[string]any{}
		for label := range labels {
			numbers := make([]float64, 0, 4)
			for _, row := range window {
				if v, ok := row.Values[label]; ok {
					if f, isNum := asFloat(v); isNum {
						numbers = append(numbers, f)
					}
				}
			}
			if len(numbers) != 4 {
				continue
			}
			switch {
			case flowLabels.has(label):
				sum := 0.0
				for _, n := range numbers {
					sum += n
				}
				values[label] = sum
			case meanLabels.has(label):
				sum := 0.0
				for _, n := range numbers {
					sum += n
				}
				values[label] = sum / 4
			default:
				values[label] = numbers[3]
			}
		}
		rows = append(rows, PeriodRow{ReportPeriod: last.ReportPeriod, Values: values})
	}
	return rows
}

// RowFilters are the Financial Datasets report_period comparison filters.
type RowFilters struct {
	Exact *time.Time
	GTE   *time.Time
	LTE   *time.Time
	GT    *time.Time
	LT    *time.Time
}

// FilterRows applies the Financial Datasets report_period comparison
// filters. Ported from statements.filter_rows.
func FilterRows(rows []PeriodRow, filters RowFilters) []PeriodRow {
	var kept []PeriodRow
	for _, row := range rows {
		if rowMatches(row.ReportPeriod, filters) {
			kept = append(kept, row)
		}
	}
	return kept
}

func rowMatches(day time.Time, filters RowFilters) bool {
	if filters.Exact != nil && !day.Equal(*filters.Exact) {
		return false
	}
	if filters.GTE != nil && day.Before(*filters.GTE) {
		return false
	}
	if filters.LTE != nil && day.After(*filters.LTE) {
		return false
	}
	if filters.GT != nil && !day.After(*filters.GT) {
		return false
	}
	if filters.LT != nil && !day.Before(*filters.LT) {
		return false
	}
	return true
}

func consecutiveQuarters(window []PeriodRow) bool {
	previous := -1
	for _, row := range window {
		ordinal := quarterOrdinal(row.ReportPeriod)
		if previous != -1 && ordinal != previous+1 {
			return false
		}
		previous = ordinal
	}
	return true
}

func sectionRoot(root map[string]any, statement string) (map[string]any, error) {
	keys, ok := statementSections[statement]
	if !ok {
		return nil, schemaDriftf("unknown statement %q", statement)
	}
	for _, key := range keys {
		if section, ok := root[key].(map[string]any); ok {
			return section, nil
		}
	}
	return nil, schemaDriftf("DefiLlama payload omitted the %s statement", statement)
}

func parseMatrix(section map[string]any, statement, period string) ([]PeriodRow, error) {
	blockValue, present := section[period]
	if !present || blockValue == nil {
		return nil, nil
	}
	block, ok := blockValue.(map[string]any)
	if !ok {
		return nil, schemaDriftf("DefiLlama %s.%s has an unknown shape", statement, period)
	}
	labels, err := stringList(section["labels"], fmt.Sprintf("DefiLlama %s labels", statement))
	if err != nil {
		return nil, err
	}
	dates, err := dateListTruncated(block["periodEnding"], fmt.Sprintf("DefiLlama %s.%s.periodEnding", statement, period))
	if err != nil {
		return nil, err
	}
	values, err := valueRowsLoose(block["values"], len(labels), len(dates), fmt.Sprintf("DefiLlama %s.%s.values", statement, period))
	if err != nil {
		return nil, err
	}
	rows := make([]PeriodRow, len(dates))
	for i, day := range dates {
		rows[i] = PeriodRow{ReportPeriod: day, Values: map[string]any{}}
	}
	for rowIndex, label := range labels {
		for column := range dates {
			if item := values[rowIndex][column]; item != nil {
				rows[column].Values[label] = item
			}
		}
	}
	childrenBlock, _ := block["children"].(map[string]any)
	childDefinitions, _ := section["children"].(map[string]any)
	if childrenBlock != nil && childDefinitions != nil {
		if err := parseChildren(childrenBlock, childDefinitions, period, dates, rows, fmt.Sprintf("DefiLlama %s.%s", statement, period)); err != nil {
			return nil, err
		}
	}
	return rows, nil
}

func parseChildren(childrenBlock, childDefinitions map[string]any, period string, dates []time.Time, rows []PeriodRow, name string) error {
	definitionsValue, ok := childDefinitions[period]
	if !ok {
		return nil
	}
	definitions, ok := definitionsValue.(map[string]any)
	if !ok {
		return nil
	}
	for parent, parentBlockValue := range childrenBlock {
		parentBlock, ok := parentBlockValue.(map[string]any)
		if !ok {
			return schemaDriftf("%s children must map parents to blocks", name)
		}
		definitionValue, ok := definitions[parent]
		if !ok || definitionValue == nil {
			continue
		}
		definition, ok := definitionValue.(map[string]any)
		if !ok {
			return schemaDriftf("%s child labels for %s must be a list", name, parent)
		}
		childLabelsValue, hasLabels := definition["labels"]
		childValuesValue, hasValues := parentBlock["values"]
		if (!hasLabels || childLabelsValue == nil) && (!hasValues || childValuesValue == nil) {
			continue
		}
		rawLabels, ok := childLabelsValue.([]any)
		if !ok {
			return schemaDriftf("%s child labels for %s must be a list", name, parent)
		}
		if _, ok := childValuesValue.([]any); !ok {
			return schemaDriftf("%s child values for %s must be a list", name, parent)
		}
		childLabels := make([]string, 0, len(rawLabels))
		for _, item := range rawLabels {
			if s, ok := item.(string); ok {
				childLabels = append(childLabels, s)
			}
		}
		childValues, err := valueRowsLoose(childValuesValue, len(childLabels), len(dates), fmt.Sprintf("%s children of %s", name, parent))
		if err != nil {
			return err
		}
		for rowIndex, label := range childLabels {
			for column := range dates {
				if item := childValues[rowIndex][column]; item != nil {
					rows[column].Values[parent+"|"+label] = item
				}
			}
		}
	}
	return nil
}

func stringList(value any, name string) ([]string, error) {
	items, ok := value.([]any)
	if !ok {
		return nil, schemaDriftf("%s must be a list", name)
	}
	labels := make([]string, 0, len(items))
	for _, item := range items {
		s, ok := item.(string)
		if !ok {
			return nil, schemaDriftf("%s must contain strings", name)
		}
		labels = append(labels, s)
	}
	return labels, nil
}

// dateListTruncated parses ISO date-time strings, keeping only the first 10
// characters (statements.py's date.fromisoformat(item[:10])).
func dateListTruncated(value any, name string) ([]time.Time, error) {
	items, ok := value.([]any)
	if !ok {
		return nil, schemaDriftf("%s must be a list", name)
	}
	dates := make([]time.Time, 0, len(items))
	for _, item := range items {
		s, ok := item.(string)
		if !ok {
			return nil, schemaDriftf("%s must contain strings", name)
		}
		if len(s) > 10 {
			s = s[:10]
		}
		day, err := time.Parse(dateLayout, s)
		if err != nil {
			return nil, schemaDriftf("%s contains a non-ISO date", name)
		}
		dates = append(dates, day)
	}
	return dates, nil
}

// valueRowsLoose returns the raw matrix rows without numeric validation
// (statements.py's own _value_rows only checks list shape).
func valueRowsLoose(value any, rowCount, columnCount int, name string) ([][]any, error) {
	items, ok := value.([]any)
	if !ok || len(items) != rowCount {
		return nil, schemaDriftf("%s row count does not match labels", name)
	}
	rows := make([][]any, len(items))
	for rowIndex, rawRow := range items {
		row, ok := rawRow.([]any)
		if !ok || len(row) != columnCount {
			return nil, schemaDriftf("%s row %d has the wrong width", name, rowIndex)
		}
		rows[rowIndex] = row
	}
	return rows, nil
}

// asFloat reports whether v is a JSON number (encoding/json decodes all
// JSON numbers into float64 when unmarshaled into `any`, and booleans
// decode to bool, so this alone matches Python's
// isinstance(v, int | float) and not isinstance(v, bool)).
func asFloat(v any) (float64, bool) {
	f, ok := v.(float64)
	return f, ok
}
